reverse_iterator.hpp

```diff
@@ -0,0 +1,224 @@
+            iterator_type base = _iter;
+            ++base;
+            return base;
+        }
+
+        template<class Iter, class unused>
+        bool reverse_iterator<Iter, unused>::operator==(const reverse_iterator &other) {
+            return (_iter == other._iter);
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused>
+        reverse_iterator<Iter, unused>::operator[](difference_type index) {
+            if constexpr (lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>){
+                return *this + index;
+            }
+            else{
+                static_assert(lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>,"Iterator must be random_access for operator []");
+            }
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused> &
+        reverse_iterator<Iter, unused>::operator-=(difference_type index) {
+            if constexpr (lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>){
+                reverse_iterator<Iter>::_iter += index;
+                return *this;
+            }
+            else{
+                static_assert(lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>,"Iterator must be random_access for operator -=");
+            }
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused> &
+        reverse_iterator<Iter, unused>::operator+=(difference_type index) {
+            if constexpr (lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>){
+                reverse_iterator<Iter>::_iter -= index;
+                return *this;
+            }
+            else{
+                static_assert(lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>,"Iterator must be random_access for operator +=");
+            }
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused>
+        reverse_iterator<Iter, unused>::operator-(difference_type index) const {
+            if constexpr (lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>){
+                reverse_iterator copy = *this;
+                return copy -= index;
+            }
+            else{
+                static_assert(lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>,"Iterator must be random_access for operator -");
+            }
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused>
+        reverse_iterator<Iter, unused>::operator+(difference_type index) const {
+            if constexpr (lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>){
```

Comment:
or you can perform assertion first then do all the remaining staff without checking condition

Reply:
actually it doesn't work like assert . It tries to compile anyways , giving compile errors I don't want to see

Reply:
ok